src/main/java/org/mockito/internal/junit/JUnitRule.java

```diff
@@ -7,88 +7,85 @@
 import org.junit.runners.model.FrameworkMethod;
 import org.junit.runners.model.Statement;
 import org.mockito.Mockito;
-import org.mockito.MockitoAnnotations;
 import org.mockito.internal.util.MockitoLogger;
 import org.mockito.junit.MockitoRule;
 
+import java.util.Collection;
+import java.util.LinkedList;
+
 /**
  * Internal implementation.
  */
 public class JUnitRule implements MockitoRule {
-	
+
+    public enum Strictness { SILENT, WARN, STRICT_STUBS }
 	private final MockitoLogger logger;
-    private final boolean silent;
+    private final Collection<MockitoTestListener> listeners = new LinkedList<MockitoTestListener>();
 
     /**
      * @param logger target for the stubbing warnings
-     * @param silent whether the rule
+     * @param silent whether the rule emits warnings
      */
     public JUnitRule(MockitoLogger logger, boolean silent) {
-		this.logger = logger;
-        this.silent = silent;
-    }
-
-	@Override
-	public Statement apply(final Statement base, FrameworkMethod method, final Object target) {
-        if (silent) {
-            return new SilentStatement(target, base);
-        } else {
-            String testName = target.getClass().getSimpleName() + "." + method.getName();
-            return new DefaultStatement(target, testName, base);
-        }
+        //TODO avoid 2 very similar constructors
+		this(logger, silent? Strictness.SILENT : Strictness.WARN);
     }
 
-    public JUnitRule silent() {
-        return new JUnitRule(logger, true);
-    }
-
-    private class SilentStatement extends Statement {
-        private final Object target;
-        private final Statement base;
-
-        public SilentStatement(Object target, Statement base) {
-            this.target = target;
-            this.base = base;
-        }
-
-        public void evaluate() throws Throwable {
-            MockitoAnnotations.initMocks(target);
-            base.evaluate();
-            Mockito.validateMockitoUsage();
+    /**
+     * @param logger target for the stubbing warnings
+     * @param strictness how strict mocking / stubbing is concerned
+     */
+    public JUnitRule(MockitoLogger logger, Strictness strictness) {
+        this.logger = logger;
+        if (strictness == Strictness.SILENT) {
```

Comment:
@szczepiq 
please replace this by an swich case e.g like this to avoid code duplication:

```java
 listeners.add(newTestListener(strictness));
```
```java
MockitoTestListener newTestListener(Strictness strictness){
  switch(strictness){
    case SILENT:
       return new SilentTestListener();
   case WARN:
      return new WarningTestListener(logger);
  case STRICT_STUBS:
     return new StrictStubsTestListener();
  }
  throw new IAE("Unsuported Strictness "+strictness);
}
```